File: EMF/go/emf/emfpublisher.go
package emf

import (
	proto "github.com/golang/protobuf/proto"
	zmq "github.com/pebbe/zmq4"
	"go.uber.org/zap"

	List "container/list"
	"math/rand"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Address Prefix for Publisher.
const PUB_TCP_PREFIX = "tcp://*:"

// Regex Pattern for Topic validation.
const TOPIC_PATTERN = "^[a-zA-Z0-9-_./]+$"

// Callback to get error code for start of EMF publisher.
// [As of now, Not being used]
type EMFStartCB func(code EMFErrorCode)

// Callback to get error code for stop of EMF publisher.
// [As of now, Not being used]
type EMFStopCB func(code EMFErrorCode)

// Error Callback for start/stop of EMF publisher.
// [As of now, Not being used]
type EMFErrorCB func(code EMFErrorCode)

//Structure represents EMFPublisher.
type EMFPublisher struct {
	port          int
	startCallback EMFStartCB
	stopCallback  EMFStopCB
	errorCallback EMFErrorCB

	publisher *zmq.Socket
	context   *zmq.Context
}

// Contructs EMFPublisher.
func GetEMFPublisher(port int, startCallback EMFStartCB, stopCallback EMFStopCB, errorCallback EMFErrorCB) *EMFPublisher {
	var instance *EMFPublisher
	instance = &EMFPublisher{}
	instance.port = port
	instance.startCallback = startCallback
	instance.stopCallback = stopCallback
	instance.errorCallback = errorCallback
	instance.context = GetInstance().getContext()

	if nil == instance.context {
		logger.Error("Context is null")
	}
	instance.publisher = nil
	InitLogger()
	return instance
}

// Starts PUB instance.
func (pubInstance *EMFPublisher) Start() EMFErrorCode {
	if nil == pubInstance.context {
		logger.Error("Context is null")
		return EMF_ERROR
	}

	if nil == pubInstance.publisher {
		var err error
		pubInstance.publisher, err = instance.context.NewSocket(zmq.PUB)
		if nil != err {
			logger.Error("Publisher Socket creation failed")
		}
		var address string = getPubSocketAddress(pubInstance.port)
		err = pubInstance.publisher.Bind(address)
		if nil != err {
			logger.Error("Error while starting publisher")
			pubInstance.publisher = nil
			return EMF_ERROR
		}
		logger.Debug("Publisher started", zap.String("address", address))
	}
	return EMF_OK
}

// Publish events on the socket for subscribers.
func (pubInstance *EMFPublisher) Publish(event Event) EMFErrorCode {
	if nil == pubInstance.publisher {
		logger.Error("Publisher is null")
		return EMF_ERROR
	}

	//marshal the protobuf event
	byteEvent, err := proto.Marshal(&event)
	if nil != err {
		logger.Error("Error occured while marshalling event")
		return EMF_ERROR
	}
	//Publish event
	result, err := pubInstance.publisher.SendBytes(byteEvent, 0)
	if nil != err {
		logger.Error("Error while publishing", zap.Int("Sent bytes", result))
		return EMF_ERROR
	}
	logger.Debug("Published without topic")
	return EMF_OK
}

// Publish events on a specific topic on socket for subscribers.
//
// Note:
// (1) Topic name should be as path format. For example:home/livingroom/
// (2) Topic name can have letters [a-z, A-z], numerics [0-9] and special characters _ - / and .
func (pubInstance *EMFPublisher) PublishOnTopic(topic string, event Event) EMFErrorCode {
	if nil == pubInstance.publisher {
		return EMF_ERROR
	}

	//validate the topic
	validTopic := sanitizeTopic(topic)
	if validTopic == "" {
		return EMF_INVALID_TOPIC
	}

	//Publish event
	result, err := pubInstance.publisher.Send(validTopic, zmq.SNDMORE)
	if nil != err {
		logger.Error("Error while publishing", zap.Int("Sent bytes", result))
		return EMF_ERROR
	}

	//marshal the protobuf event
	byteEvent, err := proto.Marshal(&event)
	if nil != err {
		return EMF_ERROR
	}
	result1, err1 := pubInstance.publisher.SendBytes(byteEvent, 0)
	if nil != err1 {
		logger.Error("Error while publishing", zap.Int("Sent bytes", result1))
		return EMF_ERROR
	}
	logger.Debug("Published on topic", zap.String("Topic", validTopic))
	return EMF_OK
}

// Publish an events on list of topics on socket for subscribers. On any of
// the topic in list, if it failed to publish event it will return
// EMF_ERROR/EMF_INVALID_TOPIC.
//
// Note:
// (1) Topic name should be as path format. For example:home/livingroom/
// (2) Topic name can have letters [a-z, A-z], numerics [0-9] and special characters _ - / and .
func (pubInstance *EMFPublisher) PublishOnTopicList(topicList List.List, event Event) EMFErrorCode {
	if topicList.Len() == 0 {
		return EMF_INVALID_TOPIC
	}
	for topic := topicList.Front(); topic != nil; topic = topic.Next() {
		result := pubInstance.PublishOnTopic(topic.Value.(string), event)
		if result != EMF_OK {
			return EMF_ERROR
		}
	}
	return EMF_OK
}

// Stops PUB instance.
func (pubInstance *EMFPublisher) Stop() EMFErrorCode {
	if nil == pubInstance.publisher {
		logger.Error("Publisher is null")
		return EMF_ERROR
	}

	// Sync close
	result := pubInstance.syncClose()
	if result == EMF_OK {
		pubInstance.publisher = nil
		logger.Debug("Publisher Stopped")
	}
	return result
}

// Get publisher port.
func (pubInstance *EMFPublisher) GetPort() int {
	return pubInstance.port
}

func getPubSocketAddress(port int) string {
	return string(PUB_TCP_PREFIX) + strconv.Itoa(port)
}

func sanitizeTopic(topic string) string {
	if topic == "" {
		return topic
	}

	result, _ := regexp.MatchString(TOPIC_PATTERN, topic)
	if false == result {
		return ""
	}

	index := strings.LastIndex(topic, "/")
	if index == -1 {
		return topic + "/"
	}

	if index != (len(topic) - 1) {
		return topic + "/"
	}
	return topic
}

func getMonitorAddress() string {
	var address string = "inproc://monitor-"
	source := rand.NewSource(time.Now().UnixNano())
	random := rand.New(source)
	return address + strconv.Itoa(random.Intn(10000))
}

func (pubInstance *EMFPublisher) syncClose() EMFErrorCode {
	var errMonitor error
	var address string = getMonitorAddress()
	errMonitor = pubInstance.publisher.Monitor(address, zmq.EVENT_CLOSED)
	if errMonitor != nil {
		logger.Info("Error in monitor")
	}
	socket, errMonitor := pubInstance.context.NewSocket(zmq.PAIR)
	if errMonitor == nil {
		errMonitor = socket.Connect(address)
		if errMonitor != nil {
			logger.Info("Pair socket connection failed")
		}
	} else {
		logger.Info("Pair socket creation failed")
	}

	//close the publisher socket
	err := pubInstance.publisher.Close()
	if nil != err {
		return EMF_ERROR
	}
	logger.Debug("Closed publisher socket")

	if nil == socket || errMonitor != nil {
		return EMF_OK
	}
	// set receive timeout
	socket.SetRcvtimeo(time.Second)
	for {
		event, addr, value, err := socket.RecvEvent(0)
		if err != nil {
			logger.Error("Error while receiving Event")
			socket.Close()
			break
		}
		logger.Debug("Event received", zap.Int("eventType", int(event)), zap.String("address", addr),
			zap.Int("Value", value))
		if event == zmq.EVENT_CLOSED {
			logger.Debug("Received EVENT_CLOSED from socket")
			socket.Close()
			break
		}
	}
	return EMF_OK
}

File: java/edgex-emf/src/main/java/org/edgexfoundry/emf/EMFPublisher.java
package org.edgexfoundry.emf;

import java.util.List;
import java.util.UUID;
import java.util.concurrent.locks.ReentrantLock;

import org.edgexfoundry.domain.core.Event;
import org.edgexfoundry.emf.protobufevent.EMFEventConverter;
import org.edgexfoundry.support.logging.client.EdgeXLogger;
import org.edgexfoundry.support.logging.client.EdgeXLoggerFactory;
import org.zeromq.ZMQ;

/**
 * This class represents the EMF Publisher.
 */
public class EMFPublisher {

    private int mPort;
    @SuppressWarnings("unused")
    private EMFCallback mCallback;

    private ZMQ.Socket mPublisher;
    private ZMQ.Context mContext;

    // Thread safety lock
    private ReentrantLock mPubLock;

    private final String PUB_TCP_PREFIX = "tcp://*:";
    private final String TOPIC_PATTERN = "[a-zA-Z0-9-_./]+";
    private final static EdgeXLogger logger = EdgeXLoggerFactory.getEdgeXLogger(EMFPublisher.class);

    /**
     * Publish data on specified port number. {@link EMFPublisher#start} API
     * should be called before publishing the message.
     *
     * @param port
     *            port for publishing message/events.
     * @param callback
     *            {@link EMFCallback}
     */
    public EMFPublisher(int port, EMFCallback callback) {
        mPort = port;
        mCallback = callback;
        mContext = EMFAPI.getInstance().getContext();
        mPubLock = new ReentrantLock(true);
    }

    // finalize method to be called by Java Garbage collector before destroying
    // this object.
    @Override
    protected void finalize() {
        logger.debug("In finalize");
        stop();
    }

    /**
     * Starts PUB instance.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode start() {
        if (null == mContext) {
            logger.error("Context is null");
            return EMFErrorCode.EMF_ERROR;
        }

        try {
            mPubLock.lock();
            if (null == mPublisher) {
                mPublisher = mContext.socket(ZMQ.PUB);
                mPublisher.bind(getSocketAddress());
            }
        } catch (Exception e) {
            logger.error("Exception while starting publisher: " + e.getMessage());
            mPublisher = null;
            return EMFErrorCode.EMF_ERROR;
        } finally {
            mPubLock.unlock();
        }
        logger.debug("Publisher started [address]: " + getSocketAddress());
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Publish events on the socket for subscribers.
     *
     * @param event
     *            {@link Event}
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode publish(Event event) {
        if (null == mPublisher || null == event) {
            logger.error("Publisher or event is null");
            return EMFErrorCode.EMF_ERROR;
        }
        byte[] byteEvent = EMFEventConverter.toProtoBuf(event);
        if (null == byteEvent) {
            logger.error("ByteEvent is null");
            return EMFErrorCode.EMF_ERROR;
        }
        boolean result = false;
        try {
            mPubLock.lock();
            if (null != mPublisher) {
                result = mPublisher.send(byteEvent);
            } else {
                return EMFErrorCode.EMF_ERROR;
            }
        } catch (Exception e) {
            logger.error("Exception while publishing: " + e.getMessage());
            return EMFErrorCode.EMF_ERROR;
        } finally {
            mPubLock.unlock();
        }
        if (false == result) {
            logger.error("Published without topic failed");
            return EMFErrorCode.EMF_ERROR;
        }
        logger.debug("Published without topic");
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Publish events on a specific topic on socket for subscribers.
     *
     * Note (1) Topic name should be as path format. For example:
     * home/livingroom/ (2) Topic name can have letters [a-z, A-z], numerics
     * [0-9] and special characters _ - . and / (3) Topic will be appended with
     * forward slash [/] in case, if application has not appended it.
     *
     * @param topic
     *            Topic on which event needs to be published.
     * @param event
     *            {@link Event}
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode publish(String topic, Event event) {
        if (null == mPublisher || null == event) {
            logger.error("Publisher or event is null");
            return EMFErrorCode.EMF_ERROR;
        }

        // validate topic
        String validTopic = validateTopic(topic);
        if (null == validTopic) {
            logger.error("Invalid topic: " + topic);
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }

        boolean result = false;
        byte[] byteEvent = EMFEventConverter.toProtoBuf(event);
        if (null == byteEvent) {
            logger.error("byteEvent is null");
            return EMFErrorCode.EMF_ERROR;
        }
        try {
            mPubLock.lock();
            if (null != mPublisher) {
                result = mPublisher.sendMore(validTopic);
                if (false == result) {
                    logger.error("SendMore failed");
                    return EMFErrorCode.EMF_ERROR;
                }
                result = mPublisher.send(byteEvent);
            } else {
                return EMFErrorCode.EMF_ERROR;
            }
        } catch (Exception e) {
            logger.error("Exception while publishing: " + e.getMessage());
            return EMFErrorCode.EMF_ERROR;
        } finally {
            mPubLock.unlock();
        }
        logger.debug("Published on topic: " + validTopic);
        if (false == result) {
            logger.error("SendMore failed");
            return EMFErrorCode.EMF_ERROR;
        }
        return EMFErrorCode.EMF_OK;
    }

    /**
     * Publish an events on list of topics on socket for subscribers. On any of
     * the topic in list, if it failed to publish event it will return
     * EMF_ERROR/EMF_INVALID_TOPIC.
     *
     * Note: (1) Topic name should be as path format. For example:
     * home/livingroom/ (2) Topic name can have letters [a-z, A-z], numerics
     * [0-9] and special characters _ - . and / (3) Topic will be appended with
     * forward slash [/] in case, if application has not appended it.
     *
     * @param topics
     *            Topic on which event needs to be published.
     * @param event
     *            {@link Event}
     * @return {@link EMFErrorCode}
     *
     */
    public EMFErrorCode publish(List<String> topics, Event event) {
        if (null == event) {
            logger.error("Event is null");
            return EMFErrorCode.EMF_ERROR;
        }

        if (null == topics) {
            logger.error("Topic list is null");
            return EMFErrorCode.EMF_INVALID_TOPIC;
        }

        EMFErrorCode result = EMFErrorCode.EMF_OK;
        for (String topic : topics) {
            result = publish(topic, event);
            if (result != EMFErrorCode.EMF_OK) {
                return result;
            }
        }
        return result;
    }

    /**
     * Stops PUB instance.
     *
     * @return {@link EMFErrorCode}
     */
    public EMFErrorCode stop() {
        if (null == mPublisher) {
            logger.error("Publisher is null");
            return EMFErrorCode.EMF_ERROR;
        }

        EMFErrorCode result = EMFErrorCode.EMF_ERROR;
        try {
            mPubLock.lock();
            if (null != mPublisher) {
                // Sync close
                result = syncClose();
            } else {
                return result;
            }
            mPublisher = null;
        } finally {
            mPubLock.unlock();
        }
        logger.debug("Publisher stopped");
        return result;
    }

    /**
     * Get the port of the publisher.
     *
     * @return port number as integer.
     */
    public int getPort() {
        return mPort;
    }

    private String validateTopic(String topic) {
        if (null == topic || topic.isEmpty()) {
            return null;
        }

        // check whether topic contains only alphabet, digits and special
        // characters _ - . and /
        if (!topic.matches(TOPIC_PATTERN)) {
            return null;
        }

        // check whether last character is forward slash or not
        // if not append and return the string
        if (!topic.endsWith("/")) {
            topic = topic + "/";
        }
        return topic;
    }

    private String getSocketAddress() {
        return PUB_TCP_PREFIX + Integer.toString(mPort);
    }

    private static String getMonitorAddress() {
        String MONITOR_PREFIX = "inproc://monitor-";
        return MONITOR_PREFIX + UUID.randomUUID().toString();
    }

    private EMFErrorCode syncClose() {
        String address = getMonitorAddress();
        boolean result = mPublisher.monitor(address, ZMQ.EVENT_CLOSED);
        if (false == result) {
            logger.info("Error in monitor");
        }

        ZMQ.Socket socket = mContext.socket(ZMQ.PAIR);
        if (socket != null) {
            result = socket.connect(address);
            logger.info("Pair socket connection result: " + result);
        } else {
            logger.info("Pair socket creation failed");
        }

        try {
            mPublisher.close();
            logger.debug("Closed publisher socket");
        } catch (Exception e) {
            logger.error("Exception while closing socket: " + e.getMessage());
            return EMFErrorCode.EMF_ERROR;
        }

        if (null == socket || false == result) {
            return EMFErrorCode.EMF_OK;
        }

        // set receive timeout
        socket.setReceiveTimeOut(1000);
        while (true) {
            ZMQ.Event event = null;
            try {
                event = ZMQ.Event.recv(socket);
            } catch (Exception e) {
                logger.info("Exception while receiving Event");
                socket.close();
                break;
            }
            if (event == null) {
                logger.info("Event is null, timeout occured");
                socket.close();
                break;
            }

            logger.debug("Event received for socket: " + event.getEvent() + "  address: "
                    + event.getAddress());
            if (ZMQ.EVENT_CLOSED == event.getEvent()) {
                logger.debug("Received EVENT_CLOSED from socket");
                socket.close();
                break;
            }
        }
        return EMFErrorCode.EMF_OK;
    }
}
